### PR TITLE
Use raytracing for distance computations

### PR DESCRIPTION
**Pull request by jorgensd**
Also refactors the coefficient packing.

### CODE REVIEW

File: cpp/SubMesh.cpp [hunk 1]
```diff
@@ -211,4 +211,29 @@ void dolfinx_contact::SubMesh::update_geometry(
...
   dolfinx_contact::update_geometry(u_sub, _mesh);
+}
+//-----------------------------------------------------------------------------------------------
+
+std::vector<std::int32_t> dolfinx_contact::SubMesh::get_submesh_tuples(
+    int i, std::span<const std::int32_t> facets) const
+{
+  assert(_mesh);
+
+  // Map (cell, facet) tuples from parent to sub mesh for candidate surface
```
Comment (SarahRo) [inline cpp/SubMesh.cpp:222]:
```suggestion
  // Map (cell, facet) tuples from parent to sub mesh
```

File: cpp/utils.cpp [hunk 2]
```diff
@@ -5,16 +5,17 @@
...
+#include "RayTracing.h"
 #include "error_handling.h"
 #include "geometric_quantities.h"
 #include <dolfinx/geometry/BoundingBoxTree.h>
 #include <dolfinx/geometry/utils.h>
-using namespace dolfinx_contact;
 
 //-----------------------------------------------------------------------------
-void dolfinx_contact::pull_back(mdspan3_t J, mdspan3_t K,
+void dolfinx_contact::pull_back(mdspan3_t J, dolfinx_contact::mdspan3_t K,
```
Comment (SarahRo) [inline cpp/utils.cpp:15]:
```suggestion
void dolfinx_contact::pull_back(dolfinx_contact::mdspan3_t J, dolfinx_contact::mdspan3_t K,
```

File: cpp/utils.h [hunk 3]
```diff
@@ -223,14 +235,401 @@ void compute_physical_points(const dolfinx::mesh::Mesh& mesh,
...
                      std::span<const std::int32_t> candidate_facets,
-                     const QuadratureRule& q_rule);
+                     const QuadratureRule& q_rule,
+                     dolfinx_contact::ContactMode mode);
+
+/// Compute the relation between two meshes (mesh_q) and
+/// (mesh_c) by computing the intersection of rays from
+/// mesh_q onto mesh_c at a specific set of quadrature
+/// points on a subset of facets. There is also a subset of
+/// facets on mesh_c we use for intersection checks.
```
Comment (SarahRo) [inline cpp/utils.h:259]:
```suggestion
/// facets on mesh_c we use for intersection checks.
/// NOTE: If the ray intersects with more than one facet 
/// in the subset of facets on mesh_c, only one of these 
/// facets is detected and it is not guaranteed to be the 
/// closest.
```

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 4]
```diff
@@ -130,11 +130,15 @@ def nitsche_unbiased(ufl_form: ufl.Form, u: _fem.Function, markers: list[_cpp.me
     # Custom assembly
     # create contact class
     with _common.Timer("~Contact: Init"):
+        mode = dolfinx_contact.cpp.ContactMode.Raytracing
+        mode = dolfinx_contact.cpp.ContactMode.ClosestPoint
```
Comment (SarahRo) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:134]: Should the mode be input to nitsche_unbiased?

Comment (jorgensd) [reply python/dolfinx_contact/unbiased/nitsche_unbiased.py:134]: Eventually. I'll first get everything working. Currently im at a debugging stage

**Pull request merged**